### PR TITLE
Intermediate replicas (part 1) + other goodies (master)

### PR DESCRIPTION
**Pull request by alanking**
Core tests passed.

### CODE REVIEW

File: lib/core/include/transport/default_transport.hpp [hunk 1]
```diff
@@ -267,6 +267,10 @@ namespace irods::experimental::io::NAMESPACE_IMPL
 
             _func(input);
 
+            if (input.openFlags | O_CREAT) {
```
Comment (korydraughn) [inline lib/core/include/transport/default_transport.hpp]: This looks like it should be `if ((input.openFlags & O_CREAT) == O_CREAT)`.

Comment (alanking) [reply lib/core/include/transport/default_transport.hpp]: Besides your being correct, this is a residual diff from an experiment during development of this change. Will delete. Thanks!

File: server/api/src/rsDataObjTrim.cpp [hunk 2]
```diff
@@ -12,172 +12,252 @@
...
-            return CAT_INSUFFICIENT_PRIVILEGE_LEVEL;
-        }
-        accessPerm = NULL;
+    unsigned long get_minimum_replica_count(const char* copies_kw) {
+        if (!copies_kw) {
+            return DEF_MIN_COPY_CNT;
+        }
+        try {
+            const auto minimum_replica_count = std::stoul(copies_kw);
+            if (minimum_replica_count <= 0) {
```
Comment (trel) [inline server/api/src/rsDataObjTrim.cpp]: Codacy Issue found: [Checking if unsigned variable 'minimum_replica_count' is less than zero.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4331469)

File: scripts/irods/test/test_chunkydevtest.py [hunk 3]
```diff
@@ -713,8 +715,9 @@ def test_large_files_from_devtest(self):
                                    irodshome + "/icmdtest/testz", 'STDOUT_SINGLELINE', "Processing lfile1")
         self.admin.assert_icommand("irsync -r i:" + irodshome + "/icmdtest/testy i:" + irodshome + "/icmdtest/testz")
         self.admin.assert_icommand("irm -vrf " + irodshome + "/icmdtest/testy")
-        self.admin.assert_icommand("iphymv -vrS " + irodsdefresource + " -R " +
-                                   self.testresc + " " + irodshome + "/icmdtest/testz", 'STDOUT_SINGLELINE', "icmdtest/testz")
+        # #4343 - skipping as iphymv does not consider hierarchies
+        #self.admin.assert_icommand("iphymv -vrS " + irodsdefresource + " -R " +
+                                   #self.testresc + " " + irodshome + "/icmdtest/testz", 'STDOUT_SINGLELINE', "icmdtest/testz")
```
Comment (trel) [inline scripts/irods/test/test_chunkydevtest.py]: Codacy Issue found: [unexpected indentation (comment)](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4331469)

File: server/api/src/rsL3FilePutSingleBuf.cpp [hunk 4]
```diff
@@ -38,3 +44,109 @@ rsL3FilePutSingleBuf( rsComm_t *rsComm, int *l1descInx,
...
+        clearKeyVal( &filePutInp.condInput );
+        return PATH_REG_NOT_ALLOWED;
+    }
+    else if ( chkType == NO_CHK_PATH_PERM ) {
+        // =-=-=-=-=-=-=-
+        filePutInp.otherFlags |= NO_CHK_PERM_FLAG; // JMC - backport 4758
+    }
+
+    filePutOut_t* put_out = 0;
+    prev_resc_hier = filePutInp.resc_hier_;
```
Comment (trel) [inline server/api/src/rsL3FilePutSingleBuf.cpp]: Codacy Issue found: [Variable 'prev_resc_hier' is assigned a value that is never used.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4331469)

File: scripts/irods/test/test_icommands_file_operations.py [hunk 5]
```diff
@@ -1631,3 +1631,64 @@ def test_writing_collection_under_slash_4030(self):
...
+        self.admin.assert_icommand(['iadmin', 'mkresc', resc_name, 'unixfilesystem', lib.get_hostname() + ':' + vault_path], 'STDOUT_SINGLELINE', resc_name)
+        file_name = 'test_iput_to_resource_with_restricted_vault_permission'
+        file_path = os.path.join(self.testing_tmp_dir, file_name)
+        lib.make_file(file_path, size)
+        logical_path = os.path.join(self.admin.session_collection, file_name) # another user's home collection
+        try:
+            self.admin.assert_icommand(['iput', '-R', resc_name, file_path], 'STDERR', 'UNIX_FILE_MKDIR_ERR')
+            self.admin.assert_icommand(['ils', '-l', file_name], 'STDERR', 'does not exist')
+            session_vault_path = self.admin.get_vault_session_path()
+            self.assertTrue(False == os.path.exists(os.path.join(session_vault_path, file_name)))
```
Comment (trel) [inline scripts/irods/test/test_icommands_file_operations.py]: Codacy Issue found: [comparison to False should be 'if cond is False:' or 'if not cond:'](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4331469)

File: server/api/src/rsDataObjPhymv.cpp [hunk 6]
```diff
@@ -17,378 +17,227 @@
...
+        return source_l1descInx;
+    }
+
+    int open_destination_replica(
+        rsComm_t* rsComm,
+        dataObjInp_t& destination_data_obj_inp) {
+        addKeyVal(&destination_data_obj_inp.condInput, REG_REPL_KW, "");
+        addKeyVal(&destination_data_obj_inp.condInput, FORCE_FLAG_KW, "");
+        destination_data_obj_inp.oprType = REPLICATE_DEST;
+        destination_data_obj_inp.oprType = PHYMV_DEST;
```
Comment (trel) [inline server/api/src/rsDataObjPhymv.cpp]: Codacy Issue found: [Variable 'destination_data_obj_inp.oprType' is reassigned a value before the old one has been used.](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4331469)

File: scripts/irods/test/test_chunkydevtest.py (hunk 3 above)
Comment (trel) [inline scripts/irods/test/test_chunkydevtest.py]: Codacy Issue found: [indentation is not a multiple of four (comment)](https://app.codacy.com/manual/trel/irods/pullRequest?prid=4331469)

**Comment by alanking** [comment]
Awaiting tests... 

All Codacy issues were existent previously, but I tried to fix them up in this last push. The last one was unclear on its issue but I can address it if we feel it is necessary.

**Comment by alanking** [comment]
Two tests failed in CI:

"Install and Run iRODS Core Tests": `test_resource_types.Test_Resource_ReplicationToTwoCompound.test_iget_with_purgec`
I ran this 15 times at the bench and it passed every time. I ran the whole test class at the bench and it still passes.

"run_tests_on_topolgy" (from resource server, no SSL): `test_all_rules.Test_AllRules.test_rulemsiServerMonPerf_r`
This also passes at the bench in an equivalent topology.

**Comment by trel** [comment]
excellent  - so the test framework still needs some work - but the code sounds solid.

one codacy complaint

**Comment by alanking** [comment]
Yes, that Codacy issue existed before but appears as a diff in the change and so is re-raising it. I can fix it if you want. It seems that it wants us to check the return value of the `read` call more carefully when copying data from one file to another in the unixfilesystem resource. Thoughts?

**Comment by trel** [comment]
Always happy to see little green checkmarks.   If it's only one more check, yes, please investigate.

There were also a couple other 'scope can be reduced' issues raised as 'possible'.

**Pull request merged**